Ensure window-start and window-hscroll remain unchanged

jamescherti opened the pull request:
This commit ensures that `markdown-toc-generate-toc` does not change the window-start or window-hscroll, addressing a common issue that can alter these settings. (I mentioned this issue in [this article](https://www.jamescherti.com/emacs-markdown-table-of-contents-update-before-save/), and the version of this pull request is better than the cod snippet in the article.)

jcs090218 commented on markdown-toc.el:
```diff
@@ -272,15 +272,34 @@ the deleted TOC region, effectively replacing it."
+                            (vertical-motion 0)
+                            (point))))
+        (set-window-start (selected-window) window-start t))
+      (set-window-hscroll (selected-window) window-hscroll))))
```
Why not use the macro `save-window-excursion`? 🤔

jamescherti replied:
Using `save-window-excursion` does not work because the buffer is modified by `markdown-toc`. This pull request restores `window-start` relative to the current line after the buffer has been modified by `markdown-toc`.

jamescherti commented:
I've updated the commit and commit message, @jcs090218.

Because markdown-toc modifies the buffer content while it is visible in a window, Emacs often realigns the display to keep the point visible, which frequently causes the window-start to jump. Using save-window-excursion does not fix this issue that this commit fixes.

This is noticeable when the Table of Contents is updated at the top of a file while the user is working further down.

**Steps to Reproduce:**

1. Open a long Markdown file (longer than the window height).
2. Ensure there is an existing TOC at the top, or a placeholder for one.
3. Scroll down to the middle of the document and place your cursor on a specific line under the table of contents.
4. Note the visual position of that line (e.g., exactly in the middle of the window).
5. Ensure the TOC is above the cursor and visible in the current window, then run `M-x markdown-toc-generate-toc`.

(I recommend reducing the window width and disabling line truncation, as this makes it easier to reproduce the issue.)

jamescherti committed:
Ensure window-start remain unchanged

Generating a table of contents involves modifying the buffer content while it is
visible in a window. When Markdown content is added or removed above the current
cursor position, Emacs often realigns the display to keep the point visible,
which frequently causes the window-start to jump. This is noticeable when the
Table of Contents is generated at the top of a file while the user is working
further down. By capturing the number of visual lines between the cursor and the
top of the window before the TOC generation and restoring that relative offset
afterward, we ensure the visual layout remains stable and prevent the text from
shifting unexpectedly under the user's eyes.

This commit prevents markdown-toc-generate-toc from disrupting the user's scroll
position by preserving the window-start relative to the cursor.
